src/session.c

```diff
@@ -689,44 +691,54 @@ CK_SLOT_ID p11prov_session_slotid(P11PROV_SESSION *session)
 
 /* returns a locked login_session if _session is not NULL */
 static CK_RV token_login(P11PROV_SESSION *session, P11PROV_URI *uri,
-                         OSSL_PASSPHRASE_CALLBACK *pw_cb, void *pw_cbarg)
+                         OSSL_PASSPHRASE_CALLBACK *pw_cb, void *pw_cbarg,
+                         struct p11prov_slot *slot)
 {
     char cb_pin[MAX_PIN_LENGTH + 1] = { 0 };
     size_t cb_pin_len = 0;
     CK_UTF8CHAR_PTR pin = NULL_PTR;
     CK_ULONG pinlen = 0;
     CK_RV ret;
 
-    if (uri) {
-        pin = (CK_UTF8CHAR_PTR)p11prov_uri_get_pin(uri);
-    }
-    if (!pin) {
-        pin = p11prov_ctx_pin(session->provctx);
-    }
-    if (pin) {
-        pinlen = strlen((const char *)pin);
-    } else if (pw_cb) {
-        const char *info = "PKCS#11 Token";
-        OSSL_PARAM params[2] = {
-            OSSL_PARAM_DEFN(OSSL_PASSPHRASE_PARAM_INFO, OSSL_PARAM_UTF8_STRING,
-                            (void *)info, sizeof(info)),
-            OSSL_PARAM_END,
-        };
-        ret = pw_cb(cb_pin, sizeof(cb_pin), &cb_pin_len, params, pw_cbarg);
-        if (ret != RET_OSSL_OK) {
-            ret = CKR_GENERAL_ERROR;
-            goto done;
+    if (!(slot->token.flags & CKF_PROTECTED_AUTHENTICATION_PATH)) {
+        if (uri) {
+            pin = (CK_UTF8CHAR_PTR)p11prov_uri_get_pin(uri);
+            if (slot->bad_pin && strcmp((char *)pin, slot->bad_pin) == 0) {
+                pin = NULL;
+            }
+        }
+        if (!pin) {
+            pin = p11prov_ctx_pin(session->provctx);
+            if (slot->bad_pin && strcmp((char *)pin, slot->bad_pin) == 0) {
+                pin = NULL;
```

Comment:
I would probably at least log some debug message in these cases as this might not be something the user is expecting when they spend time debugging why the pin they provided is not being used.

Reply:
Good idea.